src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1571,32 +960,87 @@ where
     ((ok_streams, err_stream), None)
 }
 
+/// Reclocks an `IntoTime` frontier stream into a `FromTime` frontier stream. This is used for the
+/// feedback edge that converts the `IntoTime` resumption frontier into the `FromTime` frontier
+/// that is used with the source's `OffsetCommiter`.
+fn reclock_resume_upper<FromTime, IntoTime>(
+    mut resume_rx: UnboundedReceiver<EventCore<IntoTime, Vec<()>>>,
+    mut reclock_follower: ReclockFollower<FromTime, IntoTime>,
+    worker_id: usize,
+    id: GlobalId,
+) -> impl futures::stream::Stream<Item = Antichain<FromTime>>
+where
+    FromTime: SourceTimestamp,
+    IntoTime: Timestamp + Lattice + Display,
+{
+    async_stream::stream!({
+        let mut resume_upper = MutableAntichain::new_bottom(Timestamp::minimum());
+        let mut prev_upper = reclock_follower.since().to_owned();
+
+        while let Some(event) = resume_rx.recv().await {
+            if let EventCore::Progress(changes) = event {
+                resume_upper.update_iter(changes);
+
+                let source_upper = loop {
+                    match reclock_follower.source_upper_at_frontier(resume_upper.frontier()) {
```

Comment:
someday I hope we can stop sharing the reclock follower between these 2 operators :(